File: ruby/026_IndependentSetOnATree.rb
# 026 - Independent Set on a Tree（★4）
# 二部グラフ　再帰関数

N = gets.to_i
G = Array.new(N + 1) { [] }
(N - 1).times do |i|
  a, b = gets.split.map(&:to_i)
  G[a] << b
  G[b] << a
end

USED = []
G01 = [[], []]
def dfs(pos = 1, parity = 0)
  return if USED[pos]
  USED[pos] = true
  G01[parity] << pos
  G[pos].each { dfs(_1, 1 - parity) }
end

dfs(1, 0)
puts G01.max_by(&:size).take(N / 2).join(" ")

File: kotlin/026_IndependentSetOnATree.kt
/**
 * 026 - Independent Set on a Tree（★4）
 * 二部グラフ　再帰関数
 */

fun solve() {
  val N = readLine()!!.toInt()
  val G = MutableList(N) { mutableListOf<Int>() }
  for (i in 0 until N - 1) {
    val (a, b) = readLine()!!.split(" ").map { it.toInt() - 1 }
    G[a].add(b)
    G[b].add(a)
  }

  val colors = MutableList<Int>(N) { -1 }
  fun dfs(pos: Int, col: Int) {
    colors[pos] = col
    for (i in G[pos]) {
      if (colors[i] != -1) continue
      dfs(i, 1 - col)
    }
  }
  dfs(0, 0)

  val red = if (colors.sum() >= N / 2) 1 else 0
  val ans = colors.mapIndexed {i, c -> if(c == red) i + 1 else null}.filterNotNull().take(N / 2)
  println(ans.joinToString(" "))
}

fun main() {
  Thread(null, ::solve, "solve", 1.shl(26)).start()
}
